### PR TITLE
Fix key error formatting and move exc code to exc.py

### LINKED ISSUE #1953: Dynamo "You can suppress this exception and fall back to eager" garbles KeyError
### 🐛 Describe the bug

Looks like

```
KeyError: 'source\n\nSet torch._dynamo.config.verbose=True for more information\n\n\nYou can suppress this exception and fall back to eager by setting:\n    torch._dynamo.config.suppress_errors = True\n'
```

relevant tb

```
Traceback (most recent call last):
  File "/data/users/ezyang/a/pytorch/test/dynamo/test_modules.py", line 775, in test_torch_function_with_closure
    run()
  File "/data/users/ezyang/a/pytorch/test/dynamo/test_modules.py", line 768, in run
    out2 = opt_foo(x)
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/eval_frame.py", line 209, in _fn
    return fn(*args, **kwargs)
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/eval_frame.py", line 325, in catch_errors
    return callback(frame, cache_size)
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/convert_frame.py", line 102, in _fn
    return fn(*args, **kwargs)
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/utils.py", line 90, in time_wrapper
    r = func(*args, **kwargs)
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/convert_frame.py", line 339, in _convert_frame_assert
    return _compile(
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/convert_frame.py", line 458, in _compile
    raise InternalTorchDynamoError() from e
torch._dynamo.exc.InternalTorchDynamoError
```

### Error logs

_No response_

### Minified repro

_No response_

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/torchdynamo/issues/1953 and moves exception formatting code from convert_frame.py to exc.py

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
Can we get a diff without the move. Hard to compare before after

File: torch/_dynamo/exc.py [hunk 1]
```diff
@@ -70,3 +74,115 @@ def unimplemented(msg: str):
...
+    if not config.suppress_errors:
+        msg += (
+            "\n\n"
+            "You can suppress this exception and fall back to eager by setting:\n"
+            "    torch._dynamo.config.suppress_errors = True\n"
+        )
+
+    old_msg = "" if len(exc.args) == 0 else exc.args[0]
+
+    if isinstance(exc, KeyError):
```
Comment (mlazos) [inline torch/_dynamo/exc.py:128]: Normally I'd listen, but this is the only change @ezyang

Comment (mlazos) [reply torch/_dynamo/exc.py:128]: Basically KeyError is special (https://github.com/python/cpython/blob/3.11/Objects/exceptions.c#L2534)

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**